eeps/eep-0072.md

```diff
@@ -0,0 +1,195 @@
+
+The code make it possible to link the error message to an external
+resource (a wiki), which contains all the required, additional,
+information about the error that would not be practical to present
+directly to the user. Here is an example of what the entry could look
+like for the above code:
+
+![Erlang Error Index Sample Entry][]
+
+Unique error codes also have the advantage to be better searchable in
+forums and chats, where the exact error message could vary, but the
+error code would be the same.
+
+Finally, error codes can be used by IDEs (e.g. via language servers)
+to match on error codes and provide contextual help. Both the [Erlang
+LS][] and the [ELP][] language server already use "unofficial" error
+codes.
+
+Format
+------
+
+An error code should be composed by two parts: an alphanumeric
+_namespace_ (three letters) and a numeric identifier (four digits),
+divided by a dash (`-`).
+
+A potential set of namespaces could look like the following:
+
+| Namespace | Description                                                     |
+|-----------|-----------------------------------------------------------------|
+| ERL       | The Erlang compiler and related tools (linter, parser, scanner) |
+| DIA       | The Dialyzer type-checker                                       |
+| ELV       | The Elvis code-style reviewer                                   |
+| ELP       | The Erlang Language Platform                                    |
+| ...       | ...                                                             |
+
+A set of potential error codes could look like:
+
+    ERL-0123
+    DIA-0009
+    ELV-0015
+    ELP-0001
+
+The exact number of characters/digits for each namespace and code is
+open for discussion, as well as the fact whether components such as
+the parser, the scanner or the `erlint` Erlang linter should have
+their own namespace.
+
+Responsibilities
+----------------
+
+The Erlang/OTP team would be ultimately responsible for maintaining a
+  list of _official_ namespaces. Each tool maintainer would then be
+  responsible to allocate specific codes to specific diagnostics.
+
+Processes
+---------
+
+Error cannot be re-used. If a tool stops emitting an error code, the
+_deprecated_ error code is still documented in the index, together
+with a deprecation notice. This is to avoid re-using a single code for
+multiple purposes.
+
+Reference Implementation
+------------------------
+
+The [ELP website][] contains a proof of concept of what an Erlang
+Error Index could look like. Ideally, such a website would live under
+the `erlang.org` domain.
+
+The website should use _Markdown_ as the primary mechanism to write
+content and it should be easily extensible by the Community.
+
+Required Changes
+================
+
+* The `erlc` compiler and the `dialyzer` type-checker would start
+  emitting error codes together with errors and warnings messags.
+* The tools could also expose a `--explain` option from the command
```

Comment:
How would `erlc` fetch the full explanation? How does `rustc` do it?

Reply:
@garazdawi I had to check. Apparently the Markdown files are actually stored in a crate which is part of the compiler itself:

https://github.com/rust-lang/rust/blob/master/compiler/rustc_error_codes/src/error_codes/

Then they have a [registry](https://github.com/rust-lang/rust/blob/master/compiler/rustc_errors/src/registry.rs) (a hashmap) which maps each error code to its corresponding description.

When you [pass](https://github.com/rust-lang/rust/blob/edbd9398dcc0e3b6ffe0dd0cd4578d37d1d36974/compiler/rustc_driver_impl/src/lib.rs#L307) the `explain` command they simply [try to fetch](https://github.com/rust-lang/rust/blob/edbd9398dcc0e3b6ffe0dd0cd4578d37d1d36974/compiler/rustc_driver_impl/src/lib.rs#L536) the description. Not very advanced.

The initial plumbing was introduced [here](https://github.com/rust-lang/rust/commit/9b9cce2316119a2ffdc9556d410e464b7542d64d).

Reply:
IMO we should remove this as it creates a dependency between what is stored online and what is embedded in the compiler. It is better if everyone just uses the webpage directly.